### PR TITLE
Fix errors in composer container page

### PR DESCRIPTION
**Pull request by jordisala1991**
## Subject

I am targeting this branch, because this is a problem for 4.0, since some blocks were removed.

Closes #1611 

## Changelog

```markdown
### Added
- Add ability to display and remove blocks that belongs to types that no longer exist on compose page.
```

### CODE REVIEW

File: src/Resources/views/BlockAdmin/compose_preview.html.twig [hunk 1]
```diff
@@ -7,21 +7,30 @@
...
+            {% set blockTitle = child.name|default(service.metadata.title)|trans({}, service.metadata.domain|default('SonataPageBundle')) %}
+            {% set blockImage = service.metadata.image ? asset(service.metadata.image) : null %}
+            {% set blockClass = service.metadata.option('class') %}
+            {% set metadataTitle = service.metadata.title|trans({}, service.metadata.domain|default('SonataPageBundle')) %}
         {% else %}
-            {% set metadata = service.blockMetadata %}
+            {% set blockTitle = child.name|trans({}, 'SonataPageBundle') %}
+            {% set blockImage = null %}
+            {% set blockClass = null %}
+            {% set metadataTitle = 'Block service not found' %}
```
Comment (jordisala1991) [inline src/Resources/views/BlockAdmin/compose_preview.html.twig]: This probably needs to be added to translations

File: src/Resources/views/BlockAdmin/compose_preview.html.twig [hunk 2]
```diff
@@ -7,21 +7,30 @@
...
+        {% set service = attribute(blockServices, child.type) ?? null %}
+
+        {% if service is not null %}
+            {% set blockTitle = child.name|default(service.metadata.title)|trans({}, service.metadata.domain|default('SonataPageBundle')) %}
+            {% set blockImage = service.metadata.image ? asset(service.metadata.image) : null %}
+            {% set blockClass = service.metadata.option('class') %}
+            {% set metadataTitle = service.metadata.title|trans({}, service.metadata.domain|default('SonataPageBundle')) %}
         {% else %}
-            {% set metadata = service.blockMetadata %}
+            {% set blockTitle = child.name|trans({}, 'SonataPageBundle') %}
```
Comment (jordisala1991) [inline src/Resources/views/BlockAdmin/compose_preview.html.twig]: On the previous code child.name was translated, but that does not make a lot of sense to me. Am I right? @VincentLanglet @eerison @JeromeEngelnAdeliom

**Comment by jordisala1991** [comment]
I am trying an alternative for: https://github.com/sonata-project/SonataPageBundle/pull/1611

Wip atm.

File: src/Resources/views/BlockAdmin/compose_preview.html.twig (hunk 2 above)
Comment (eerison) [reply src/Resources/views/BlockAdmin/compose_preview.html.twig]: Translate child.name does not make sense! because it is a dynamic data from db

**Comment by jordisala1991** [comment]
This is a bit different implementation than #1611, 

@eerison @JeromeEngelnAdeliom can you confirm this also works? I tried to change the minimum code as possible.

File: src/Admin/BlockAdmin.php [hunk 3]
```diff
@@ -172,25 +172,25 @@ protected function configureFormFields(FormMapper $form): void
...
-            $this->configureBlockFields($form, $block);
-
             if (false !== $isComposer) {
                 $form->add('enabled', HiddenType::class, ['data' => true]);
             } else {
                 $form->add('enabled');
             }
 
             $form->end();
+
```
Comment (jordisala1991) [inline src/Admin/BlockAdmin.php:182]: Is better this way, this looks more similar to what we do on shared blocks.

File: src/Admin/BlockAdmin.php [hunk 4]
```diff
@@ -172,25 +172,25 @@ protected function configureFormFields(FormMapper $form): void
...
+            $this->configureBlockFields($form, $block);
+
+            $form->end();
         } else {
             $form
                 ->with('options', $optionsGroupOptions)
-                ->add('type', ServiceListType::class, ['context' => 'sonata_page_bundle'])
-                ->add('enabled')
-                ->add('position', IntegerType::class)
+                    ->add('type', ServiceListType::class, ['context' => 'sonata_page_bundle'])
```
Comment (jordisala1991) [inline src/Admin/BlockAdmin.php:191]: missing indentation. better to code readability.

File: src/Resources/views/BlockAdmin/compose_preview.html.twig [hunk 5]
```diff
@@ -4,29 +4,51 @@
...
+        href="{{ blockAdmin.generateUrl('edit', { 'id': child.id, 'composer': true }) }}"
     >
-        {% set service = attribute(blockServices, child.type) %}
-        {% if service.metadata is defined %}
-            {% set metadata = service.metadata %}
+    {% else %}
+    <div class="page-composer__container__child__error">
+    {% endif %}
+        {% set service = attribute(blockServices, child.type) ?? null %}
+
```
Comment (jordisala1991) [inline src/Resources/views/BlockAdmin/compose_preview.html.twig]:
```suggestion
```

File: src/Resources/views/BlockAdmin/compose_preview.html.twig [hunk 6]
```diff
@@ -4,29 +4,49 @@
     data-block-enabled="{{ child.enabled|default('0') }}"
     data-block-type="{{ child.type }}"
 >
-    <a class="page-composer__container__child__edit"
-       href="{{ blockAdmin.generateUrl('edit', { 'id': child.id, 'composer': true }) }}"
+    {% set service = attribute(blockServices, child.type) ?? null %}
+
+    {% if service %}
```
Comment (jordisala1991) [inline src/Resources/views/BlockAdmin/compose_preview.html.twig:9]:
```suggestion
    {% if service is not null %}
```

File: src/Resources/views/BlockAdmin/compose_preview.html.twig [hunk 7]
```diff
@@ -4,29 +4,49 @@
...
         {% endif %}
-        <small>{{ metadata.title|trans({}, metadata.domain|default('SonataPageBundle')) }}</small>
-        <span class="page-composer__container__child__toggle">
-            <i class="fa fa-chevron-down"></i>
-            <i class="fa fa-chevron-up"></i>
-        </span>
+
+        <small>{{ metadataTitle }}</small>
+
+        {% if service %}
```
Comment (jordisala1991) [inline src/Resources/views/BlockAdmin/compose_preview.html.twig:39]:
```suggestion
        {% if service is not null %}
```

**Review by VincentLanglet — APPROVED** [review]

**Comment by VincentLanglet** [comment]
Ready to merge @jordisala1991 ?

**Pull request merged**